core/server/proxy/src/main/java/alluxio/proxy/s3/S3Handler.java

```diff
@@ -0,0 +1,518 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.proxy.s3;
+
+import alluxio.AlluxioURI;
+import alluxio.client.file.FileSystem;
+import alluxio.conf.Configuration;
+import alluxio.conf.PropertyKey;
+import alluxio.grpc.Bits;
+import alluxio.grpc.CreateDirectoryPOptions;
+import alluxio.grpc.PMode;
+import alluxio.grpc.XAttrPropagationStrategy;
+import alluxio.master.audit.AsyncUserAccessAuditLogWriter;
+import alluxio.util.CommonUtils;
+import alluxio.util.ThreadUtils;
+import alluxio.web.ProxyWebServer;
+
+import com.google.common.base.Stopwatch;
+import org.eclipse.jetty.server.Request;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URLDecoder;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletInputStream;
+import javax.servlet.ServletOutputStream;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.core.MultivaluedMap;
+import javax.ws.rs.core.Response;
+
+/**
+ *
+ */
+public class S3Handler {
+  public static final boolean BUCKET_NAMING_RESTRICTION_ENABLED =
+          Configuration.getBoolean(PropertyKey.PROXY_S3_BUCKET_NAMING_RESTRICTIONS_ENABLED);
+  public static final int MAX_HEADER_METADATA_SIZE =
+          (int) Configuration.getBytes(PropertyKey.PROXY_S3_METADATA_HEADER_MAX_SIZE);
+  public static final boolean MULTIPART_CLEANER_ENABLED =
+          Configuration.getBoolean(PropertyKey.PROXY_S3_MULTIPART_UPLOAD_CLEANER_ENABLED);
+  // https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+  // - Undocumented edge-case, no adjacent periods with hyphens, i.e: '.-' or '-.'
+  public static final Pattern BUCKET_ADJACENT_DOTS_DASHES_PATTERN = Pattern.compile("([-\\.]{2})");
+  public static final Pattern BUCKET_INVALIDATION_PREFIX_PATTERN = Pattern.compile("^xn--.*");
+  public static final Pattern BUCKET_INVALID_SUFFIX_PATTERN = Pattern.compile(".*-s3alias$");
+  public static final Pattern BUCKET_VALID_NAME_PATTERN =
+          Pattern.compile("[a-z0-9][a-z0-9\\.-]{1,61}[a-z0-9]");
+  public static final Pattern BASE_PATH_PATTERN =
```

Comment:
I see nowhere to use this Pattern, will the customer access "/api/v1/s3" to get the bucket list?

Reply:
removed this unused pattern, and yes, this is the path to get the bucket list same as before:
bucket will be empty str "" and in creating s3 task it will be :
in S3BucketTask.Factory.create:
if (StringUtils.isEmpty(handler.getBucket())) {
            return new ListBucketsTask(handler, OpType.ListBuckets);
-> it will be routed to listbuckets

Reply:
got it